workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs

```diff
@@ -0,0 +1,161 @@
+                (GameObjectComponentDispatcherBase) Activator.CreateInstance(type)));
+        }
+
+        private void GenerateComponentGroups()
+        {
+            foreach (var gameObjectComponentDispatcher in GameObjectComponentDispatchers)
+            {
+                gameObjectComponentDispatcher.ComponentAddedComponentGroup =
+                    GetComponentGroup(gameObjectComponentDispatcher.ComponentAddedComponentTypes);
+                gameObjectComponentDispatcher.ComponentRemovedComponentGroup =
+                    GetComponentGroup(gameObjectComponentDispatcher.ComponentRemovedComponentTypes);
+                gameObjectComponentDispatcher.AuthoritiesChangedComponentGroup =
+                    GetComponentGroup(gameObjectComponentDispatcher.AuthoritiesChangedComponentTypes);
+                if (gameObjectComponentDispatcher.ComponentsUpdatedComponentTypes.Length > 0)
+                {
+                    gameObjectComponentDispatcher.ComponentsUpdatedComponentGroup =
+                        GetComponentGroup(gameObjectComponentDispatcher.ComponentsUpdatedComponentTypes);
+                }
+
+                if (gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays.Length > 0)
+                {
+                    gameObjectComponentDispatcher.EventsReceivedComponentGroups =
+                        new ComponentGroup[gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays.Length];
+                    for (var i = 0; i < gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays.Length; i++)
+                    {
+                        gameObjectComponentDispatcher.EventsReceivedComponentGroups[i] =
+                            GetComponentGroup(gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays[i]);
+                    }
+                }
+
+                if (gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays.Length > 0)
+                {
+                    gameObjectComponentDispatcher.CommandRequestsComponentGroups =
+                        new ComponentGroup[gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays.Length];
+                    for (var i = 0; i < gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays.Length; i++)
+                    {
+                        gameObjectComponentDispatcher.CommandRequestsComponentGroups[i] =
+                            GetComponentGroup(gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays[i]);
+                    }
+                }
+            }
+        }
+
+        protected override void OnUpdate()
+        {
+            foreach (var gameObjectComponentDispatcher in GameObjectComponentDispatchers)
+            {
+                gameObjectComponentDispatcher.InvokeOnAddComponentLifecycleCallbacks(this);
+                gameObjectComponentDispatcher.InvokeOnRemoveComponentLifecycleCallbacks(this);
```

Comment:
As written it's a problem if it is on the entity, then gets removed and readded in the same tick right? Because then we initially call the BehaviourManager's AddComponent() method with a component that's already on it?

Reply:
It's best to rework this to not assume anything about ordering. Use the component added and removed components to denote an entity that needs to be checked, then explicitly check if the component is there or not probably.

Reply:
The GDK core should ensure that a component can't be simultaneously added or removed.